### PR TITLE
Add waiting room event

### PR DESCRIPTION
**Pull request by broswen**
Add support for managing Waiting Room Events in Terraform.

`cloudflare_waiting_room_event` depends on a `cloudflare_waiting_room` to be created. Added into the Terraform template to create the acceptance test resources.

This is my first Terraform contribution so please let me know if there are things that should be done differently.

### CODE REVIEW

File: cloudflare/resource_cloudflare_waiting_room_event.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+func expandWaitingRoomEvent(d *schema.ResourceData) (cloudflare.WaitingRoomEvent, error) {
+	var disableSessionRenewal *bool
+	switch b := d.Get("disable_session_renewal").(type) {
+	case bool:
+		disableSessionRenewal = &b
+	case nil:
+		disableSessionRenewal = nil
+	}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_waiting_room_event.go:39]: i think we can swap this for a `d.GetOkExists` and the value can use `cloudflare.BoolPtr` to take the boolean and return a *bool.

File: cloudflare/resource_cloudflare_waiting_room_event.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	eventStartTime, err := time.Parse(time.RFC3339, d.Get("event_start_time").(string))
+	if err != nil {
+		return cloudflare.WaitingRoomEvent{}, err
+	}
+	eventEndTime, err := time.Parse(time.RFC3339, d.Get("event_end_time").(string))
+	if err != nil {
+		return cloudflare.WaitingRoomEvent{}, err
+	}
+	prequeueStartTime := time.Time{}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_waiting_room_event.go:48]:
```suggestion

	prequeueStartTime := time.Time{}
```

File: cloudflare/resource_cloudflare_waiting_room_event.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	if t, ok := d.GetOk("prequeue_start_time"); ok {
+		prequeueStartTime, err = time.Parse(time.RFC3339, t.(string))
+		if err != nil {
+			return cloudflare.WaitingRoomEvent{}, err
+		}
+	}
+
+	return cloudflare.WaitingRoomEvent{
+		ID:                    d.Id(),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_waiting_room_event.go:57]: i don't think this is going to always work like you intend it to. `d.Id()` is reliant on `d.SetId()` which you're calling later in the Create.

i think we can get rid of this and just set the `d.SetId()` to the ID returned from `waitingRoomEvent`.

File: cloudflare/resource_cloudflare_waiting_room_event.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		name := d.Get("name").(string)
+		return fmt.Errorf("error building waiting room event %q: %w", name, err)
+	}
+
+	waitingRoomEvent, err := client.CreateWaitingRoomEvent(context.Background(), zoneID, waitingRoomID, newWaitingRoomEvent)
+
+	if err != nil {
+		name := d.Get("name").(string)
+		return fmt.Errorf("error creating waiting room event %q: %w", name, err)
+	}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_waiting_room_event.go:90]:
```suggestion
	if err != nil {
		return fmt.Errorf("error creating waiting room event %q: %w", d.Get("name").(string), err)
	}
```

File: cloudflare/resource_cloudflare_waiting_room_event.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	d.Set("suspended", waitingRoomEvent.Suspended)
+	if waitingRoomEvent.TotalActiveUsers != 0 {
+		d.Set("total_active_users", waitingRoomEvent.TotalActiveUsers)
+	}
+	if waitingRoomEvent.NewUsersPerMinute != 0 {
+		d.Set("new_users_per_minute", waitingRoomEvent.NewUsersPerMinute)
+	}
+	if waitingRoomEvent.CustomPageHTML != "" {
+		d.Set("custom_page_html", waitingRoomEvent.CustomPageHTML)
+	}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_waiting_room_event.go:135]:
```suggestion
	if !waitingRoomEvent.PrequeueStartTime.IsZero() {
		d.Set("prequeue_start_time", waitingRoomEvent.PrequeueStartTime.Format(time.RFC3339))
	}
	
	if waitingRoomEvent.Description != "" {
		d.Set("description", waitingRoomEvent.Description)
	}
	
	if waitingRoomEvent.QueueingMethod != "" {
		d.Set("queueing_method", waitingRoomEvent.QueueingMethod)
	}
	
	d.Set("shuffle_at_event_start", waitingRoomEvent.ShuffleAtEventStart)
	d.Set("suspended", waitingRoomEvent.Suspended)
	
	if waitingRoomEvent.TotalActiveUsers != 0 {
		d.Set("total_active_users", waitingRoomEvent.TotalActiveUsers)
	}
	
	if waitingRoomEvent.NewUsersPerMinute != 0 {
		d.Set("new_users_per_minute", waitingRoomEvent.NewUsersPerMinute)
	}
	
	if waitingRoomEvent.CustomPageHTML != "" {
		d.Set("custom_page_html", waitingRoomEvent.CustomPageHTML)
	}
```

File: cloudflare/resource_cloudflare_waiting_room_event.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+func resourceCloudflareWaitingRoomEventUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	waitingRoomID := d.Get("waiting_room_id").(string)
+	zoneID := d.Get("zone_id").(string)
+
+	waitingRoomEvent, err := expandWaitingRoomEvent(d)
+	if err != nil {
+		name := d.Get("name").(string)
+		return fmt.Errorf("error building waiting room event %q: %w", name, err)
+	}
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_waiting_room_event.go:152]:
```suggestion
		return fmt.Errorf("error building waiting room event %q: %w", d.Get("name").(string), err)
	}
```

File: website/docs/r/waiting_room_event.html.markdown [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    event_start_time     = "2006-01-02T15:04:05Z"
+    event_end_time       = "2006-01-02T20:04:05Z"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `zone_id` - (Required) The DNS zone ID to apply to.
```
Comment (jacobbednarz) [inline website/docs/r/waiting_room_event.html.markdown:29]:
```suggestion
* `zone_id` - (Required) The zone ID to apply to.
```

File: website/docs/r/waiting_room_event.html.markdown [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+* `event_end_time` (Required) ISO 8601 timestamp that marks the end of the event.
+* `total_active_users` - (Optional) The total number of active user sessions on the route at a point in time.
+* `new_users_per_minute` - (Optional) The number of new users that will be let into the route every minute.
+* `custom_page_html` - (Optional) This a templated html file that will be rendered at the edge.
+* `queueing_method` - (Optional) The queueing method to be used by the waiting room during the event. Default: fifo.
+* `shuffle_at_event_start` - (Optional) Users in the prequeue will be shuffled randomly at the `event_start_time`. Requires that `prequeue_start_time` is not null. Default: false.
+* `disable_session_renewal` - (Optional) Disables automatic renewal of session cookies. Default: false.
+* `prequeue_start_time` - (Optional) ISO 8601 timestamp that marks when to begin queueing all users before the event starts. Must occur at least 5 minutes before event_start_time. Default: null.
```
Comment (jacobbednarz) [inline website/docs/r/waiting_room_event.html.markdown:40]: Terraform does not have the concept of "null" in schemas. We can probably omit the default bit.

File: website/docs/r/waiting_room_event.html.markdown [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+* `zone_id` - (Required) The DNS zone ID to apply to.
+* `waiting_room_id` - (Required) The Waiting Room ID the event should apply to.
+* `name` - (Required) A unique name to identify the event. Only alphanumeric characters, hyphens, and underscores are allowed.
+* `event_start_time` (Required) ISO 8601 timestamp that marks the start of the event. At this time, queued users will be processed with the event's configuration. Must occur at least 1 minute before event_end_time.
+* `event_end_time` (Required) ISO 8601 timestamp that marks the end of the event.
+* `total_active_users` - (Optional) The total number of active user sessions on the route at a point in time.
+* `new_users_per_minute` - (Optional) The number of new users that will be let into the route every minute.
+* `custom_page_html` - (Optional) This a templated html file that will be rendered at the edge.
+* `queueing_method` - (Optional) The queueing method to be used by the waiting room during the event. Default: fifo.
+* `shuffle_at_event_start` - (Optional) Users in the prequeue will be shuffled randomly at the `event_start_time`. Requires that `prequeue_start_time` is not null. Default: false.
```
Comment (jacobbednarz) [inline website/docs/r/waiting_room_event.html.markdown:38]: The default here does not match the schema default.

File: website/docs/r/waiting_room_event.html.markdown [hunk 10]
```diff
@@ -0,0 +1,63 @@
+---
```
Comment (jacobbednarz) [inline website/docs/r/waiting_room_event.html.markdown:1]: we also need to update https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/website/cloudflare.erb to include the new resource in the sidebar.

File: cloudflare/resource_cloudflare_waiting_room_event.go (hunk 1 above)
Comment (broswen) [reply cloudflare/resource_cloudflare_waiting_room_event.go]: It looks like `d.GetOkExists` is deprecated and might be removed. Would just using  `d.GetOk` be sufficient?

**Comment by jacobbednarz** [comment]
acceptance tests are passing locally

```
TF_ACC=1 go test $(go list ./...) -v -run "^TestAccCloudflareWaitingRoomEvent_Create" -count 1 -parallel 1 -timeout 120m -parallel 1
?   	github.com/cloudflare/terraform-provider-cloudflare	[no test files]
=== RUN   TestAccCloudflareWaitingRoomEvent_Create
=== PAUSE TestAccCloudflareWaitingRoomEvent_Create
=== CONT  TestAccCloudflareWaitingRoomEvent_Create
--- PASS: TestAccCloudflareWaitingRoomEvent_Create (12.98s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/cloudflare	13.359s
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/changelog-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/maintainer-only-file-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/tools/cmd/tf-log-check	[no test files]
?   	github.com/cloudflare/terraform-provider-cloudflare/version	[no test files]
```

thank you for the solid effort here @broswen

**Pull request merged**